go/libraries/doltcore/merge/merge.go

```diff
@@ -51,99 +51,101 @@ func NewMerger(ctx context.Context, root, mergeRoot, ancRoot *doltdb.RootValue,
 // MergeTable merges schema and table data for the table tblName.
 func (merger *Merger) MergeTable(ctx context.Context, tblName string, sess *editor.TableEditSession) (*doltdb.Table, *MergeStats, error) {
 	tbl, ok, err := merger.root.GetTable(ctx, tblName)
-
 	if err != nil {
 		return nil, nil, err
 	}
 
 	var h hash.Hash
 	if ok {
 		h, err = tbl.HashOf()
-
 		if err != nil {
 			return nil, nil, err
 		}
 	}
 
 	mergeTbl, mergeOk, err := merger.mergeRoot.GetTable(ctx, tblName)
-
 	if err != nil {
 		return nil, nil, err
 	}
 
 	var mh hash.Hash
 	if mergeOk {
 		mh, err = mergeTbl.HashOf()
-
 		if err != nil {
 			return nil, nil, err
 		}
 	}
 
 	ancTbl, ancOk, err := merger.ancRoot.GetTable(ctx, tblName)
-
 	if err != nil {
 		return nil, nil, err
 	}
 
 	var anch hash.Hash
+	var ancTblSchema schema.Schema
 	if ancOk {
 		anch, err = ancTbl.HashOf()
-
+		if err != nil {
+			return nil, nil, err
+		}
+		ancTblSchema, err = ancTbl.GetSchema(ctx)
 		if err != nil {
 			return nil, nil, err
 		}
 	}
 
-	if ok && mergeOk && h == mh {
-		return tbl, &MergeStats{Operation: TableUnmodified}, nil
-	}
-
-	if !ancOk {
-		if mergeOk && ok {
-			return nil, nil, ErrSameTblAddedTwice
-		} else if ok {
-			return tbl, &MergeStats{Operation: TableUnmodified}, nil
+	{ // short-circuit logic
+		if ancOk && schema.IsKeyless(ancTblSchema) {
+			if ok && mergeOk && ancOk && h == mh && h == anch {
+				return tbl, &MergeStats{Operation: TableUnmodified}, nil
+			}
 		} else {
-			return mergeTbl, &MergeStats{Operation: TableAdded}, nil
+			if ok && mergeOk && h == mh {
+				return tbl, &MergeStats{Operation: TableUnmodified}, nil
+			}
 		}
-	}
 
-	if h == anch {
-		ms := MergeStats{Operation: TableModified}
-		if h != mh {
-			ms, err = calcTableMergeStats(ctx, tbl, mergeTbl)
+		if !ancOk {
+			if mergeOk && ok {
+				return nil, nil, ErrSameTblAddedTwice
+			} else if ok {
+				// fast-forward
+				return tbl, &MergeStats{Operation: TableUnmodified}, nil
+			} else {
+				// fast-forward
+				return mergeTbl, &MergeStats{Operation: TableAdded}, nil
+			}
 		}
-		// force load the table editor since this counts as a change
-		_, err := sess.GetTableEditor(ctx, tblName, nil)
-		if err != nil {
-			return nil, nil, err
+
+		if h == anch {
+			// fast-forward
+			ms := MergeStats{Operation: TableModified}
+			if h != mh {
+				ms, err = calcTableMergeStats(ctx, tbl, mergeTbl)
+			}
+			// force load the table editor since this counts as a change
```

Comment:
What's the point of doing this? What err do we see in what circumstances?

Reply:
i have no idea...

Reply:
this was an existing line, something to do with merging foreign keys